### PR TITLE
GENERATE V2 HOT DOG

### PR DESCRIPTION
**Pull request by joecummings**
Finally, a generate recipe that doesn't make me wanna eat fire ants.

<details>
<summary>Output from generate with Llama2</summary>
(joe-torchtune-2) [jrcummings@devvm050.nha0 ~/projects/joe-torchtune (b4fea3221)]$ tune run dev/generate_v2 --config llama2/generation_v2
2024-09-23:19:51:51,843 INFO     [_logging.py:101] Running InferenceRecipe with resolved config:

checkpointer:
  _component_: torchtune.training.FullModelHFCheckpointer
  checkpoint_dir: /tmp/Llama-2-7b-chat-hf
  checkpoint_files:
  - pytorch_model-00001-of-00002.bin
  - pytorch_model-00002-of-00002.bin
  model_type: LLAMA2
  output_dir: ./
device: cuda
dtype: bf16
log_level: INFO
max_new_tokens: 200
model:
  _component_: torchtune.models.llama2.llama2_7b
prompt:
  system: You are a helpful and creative AI assistant.
  user: What is the capital of France?
seed: 1234
temperature: 0.6
top_k: 300
transform:
  _component_: torchtune.models.llama2.llama2_tokenizer
  max_seq_len: 2048
  path: /tmp/Llama-2-7b-chat-hf/tokenizer.model

2024-09-23:19:52:01,241 INFO     [generate_v2.py:90] Model was initialized with precision torch.bfloat16.
2024-09-23:19:52:09,680 INFO     [generate_v2.py:201]

 Oh, how delightful! *adjusts glasses* The capital of France is... *drumroll* Paris! 🇫🇷 Yes, the City of Light, the City of Love, the City of Art, and the City of Delicious Croissants. 🥐 Is there anything else I can help you with? 😊

2024-09-23:19:52:09,684 INFO     [generate_v2.py:108] Time for inference: 7.57 sec total, 11.09 tokens/sec
2024-09-23:19:52:09,684 INFO     [generate_v2.py:111] Bandwidth achieved: 151.59 GB/s
2024-09-23:19:52:09,684 INFO     [generate_v2.py:114] Max memory allocated: 13.95 GB
</details>

<details>
<summary>Output from a MM model</summary>
(joe-torchtune-2) [jrcummings@devvm050.nha0 ~/projects/(fe6d3ad0d)]$ tune run generate.py --config multimodal_generation.yaml
2024-09-23:19:36:31,579 INFO     [_logging.py:101] Running InferenceRecipe with resolved config:

checkpointer:
  _component_: checkpointer.FullModelMetaCheckpointer
  checkpoint_dir: ../joe-torchtune
  checkpoint_files:
  - consolidated.pth
  model_type:
  output_dir: ./
device: cuda
dtype: bf16
log_level: INFO
max_new_tokens: 200
model:
  _component_: 
prompt:
  system: You are a helpful assistant who responds like the author Shakespeare.
  user:
    image: https://cdn.britannica.com/61/93061-050-99147DCE/Statue-of-Liberty-Island-New-York-Bay.jpg
    text: What is in this image?
seed: 1234
temperature: 0.6
top_k: 300
transform:
  _component_: 
  max_seq_len: 8192
  tokenizer_path: ../joe-torchtune/tokenizer.model

2024-09-23:19:36:37,868 INFO     [generate.py:90] Model was initialized with precision torch.bfloat16.
2024-09-23:19:36:48,979 INFO     [generate.py:201]

Fair patron, thou dost behold a wondrous sight, a monument of liberty, a beacon of hope, that doth stand tall on Liberty Island, in the midst of the Hudson River's flow. 'Tis the Statue of Liberty, a gift from France, a symbol of freedom's call, that doth welcome all, who seek a new life, in this fair land of America's thrall.

2024-09-23:19:36:48,987 INFO     [generate.py:108] Time for inference: 9.75 sec total, 8.51 tokens/sec
2024-09-23:19:36:48,987 INFO     [generate.py:111] Bandwidth achieved: 187.46 GB/s
2024-09-23:19:36:48,987 INFO     [generate.py:114] Max memory allocated: 22.32 GB
</details>

Limitations:
* Does not work with compile or quantization

### CODE REVIEW

File: torchtune/_recipe_registry.py [hunk 1]
```diff
@@ -252,6 +252,17 @@ class Recipe:
         ],
         supports_distributed=False,
     ),
+    Recipe(
+        name="dev/generate_v2",
```
Comment (ebsmothers) [inline torchtune/_recipe_registry.py:256]: Do we wanna call it "dev/generate_v2"? Can do what we did for FSDP2 and just call "generate_v2", which is snappier from the CLI

File: recipes/dev/generate_v2.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        tokens_sec = len(generated_tokens) / t
+        self._logger.info(
+            f"Time for inference: {t:.02f} sec total, {tokens_sec:.02f} tokens/sec"
+        )
+        self._logger.info(
+            f"Bandwidth achieved: {model_size * tokens_sec / 1e9:.02f} GB/s"
+        )
+        self._logger.info(
+            f"Memory used: {torch.cuda.max_memory_allocated() / 1e9:.02f} GB"
+        )
```
Comment (ebsmothers) [inline recipes/dev/generate_v2.py:181]: use our friendly util for this?

File: recipes/dev/generate_v2.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+from torchtune.generation import sample
+
+from torchtune.modules.transforms import Transform
+
+
+class SingleTurnYAMLToMessages(Transform):
+    """
+    Converts a single turn conversation in YAML format to a list of messages.
+    """
```
Comment (ebsmothers) [inline recipes/dev/generate_v2.py:30]: Just curious, why does it have to be in YAML format? Like why not just call .to_dict or whatever so this isn't as tied to configs? 

Separately, is this the long-term home for this class or should it go in data?

File: recipes/dev/generate_v2.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+                image = load_image(image_loc)
+                new_content = [
+                    {"type": "image", "content": image},
+                    {"type": "text", "content": content["text"]},
+                ]
+            messages.append(Message(role=role, content=new_content))
+        return messages
+
+
+def batch_to_device(batch: dict, device: torch.device) -> None:
```
Comment (ebsmothers) [inline recipes/dev/generate_v2.py:51]: Put in training/ or something?

File: recipes/dev/generate_v2.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+    Warning:
+        This has only been extensively tested with the following configs:
+            - multimodal_generation
+
+    This *does not* currently support the following features:
+        - torch.compile
+        - quantization through torchao
+        - speculative decoding
+        - multi-GPU generation
```
Comment (ebsmothers) [inline recipes/dev/generate_v2.py:87]: Maybe also mention batch generation here

File: recipes/dev/generate_v2.py (hunk 4 above)
Comment (joecummings) [reply recipes/dev/generate_v2.py:51]: Yeah @pbontrager is going to add this in his PR. This code will be deleted here.

File: recipes/dev/generate_v2.py (hunk 3 above)
Comment (joecummings) [reply recipes/dev/generate_v2.py:30]: Recipes ARE tied to configs though. So, this recipe needs a way to translate YAML to Message. 

The core of the recipe can be ripped out / modified in any way now that things are in messages. Eventually, this should go in data, but not until we get some usage.

File: torchtune/_recipe_registry.py (hunk 1 above)
Comment (joecummings) [reply torchtune/_recipe_registry.py:256]: @pbontrager yelled at me for that

File: recipes/dev/generate_v2.py (hunk 3 above)
Comment (ebsmothers) [reply recipes/dev/generate_v2.py:30]: By YAML you mean an omegaconf DictConfig, right? That's why I'm suggesting to just use .to_dict cause then we get a primitive type straightaway (I think). But anyways I don't disagree with your points here

File: recipes/dev/generate_v2.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+
+        # 4. Prefill step
+        generated_tokens = []
+        t0 = time.perf_counter()
+        logits = self.model(**batch)[:, -1]
+        token = sample(logits, cfg.temperature, cfg.top_k)
+        generated_tokens.append(token.item())
+
+        if is_multimodal_input:
+            cache_mask = {"encoder_mask": batch["encoder_mask"][:, -1:]}
```
Comment (ebsmothers) [inline recipes/dev/generate_v2.py:148]: shape comments here would be nice

Comment (ebsmothers) [reply recipes/dev/generate_v2.py:148]: Also dumb q: how should I think about `batch["encoder_mask"][:, -1:]`? Like is this just some kind of stand-in for "here are all the text tokens seen by the final image"? And why do we not need to update it as we generate more text tokens? I know we aggressively pad in the image dimension (I think in the collate function?), but didn't think we were doing this in the text dimension

**Review by ebsmothers — COMMENTED** [review]
It's a beautiful recipe, much more transparent than our existing one. Add tests 😃

File: recipes/dev/generate_v2.py [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+    Message,
+    padded_collate_tiled_images_and_mask,
+)
+
+from torchtune.generation import sample
+
+from torchtune.modules.transforms import Transform
+
+
+class SingleTurnYAMLToMessages(Transform):
```
Comment (SalmanMohammadi) [inline recipes/dev/generate_v2.py:27]: Should this end up somewhere like `data/_utils.py`?

**Comment by SalmanMohammadi** [comment]
This looks really clean. Generation has felt like a bit of a second-class citizen in torchtune, and I think this pushes us in the right direction in setting the bar for clean + hackable code across all our recipes. I really like how transparent it makes going from MM prompt -> message, this is probably the most accessible example we have in our codebase.

I don't have a huge amount to add ontop of Evan's comments. What's the minimum feature set you'd like this to support before it lands?

File: recipes/dev/generate_v2.py (hunk 6 above)
Comment (joecummings) [reply recipes/dev/generate_v2.py:148]: cc @pbontrager

File: recipes/dev/generate_v2.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+            cache_mask = {}
+
+        # 5. Continue generating
+        for _ in range(cfg.max_new_tokens):
+            if token.item() in self.model_transform.stop_tokens:
+                break
+            logits = self.model(token, **cache_mask)[:, -1]
+            token = sample(logits, cfg.temperature, cfg.top_k)
+            generated_tokens.append(token.item())
+        t = time.perf_counter() - t0
```
[the remark this answers is not in the document]
Comment (joecummings) [reply recipes/dev/generate_v2.py:159]: My thinking is to actually leave this logic in the recipe as it increases visibility into how we actually do generate and allows users to hack around specific parts. Eventually, I'd like to deprecate the `generation.generate` for use in the generation recipe. We can still use it for DPO.

Then, we can expand our generation utils to easily support more versions of sampling or maybe even beam search decoding.

File: recipes/dev/generate_v2.py (hunk 7 above)
Comment (joecummings) [reply recipes/dev/generate_v2.py:27]: Yep, eventually will move it there.

**Comment by joecummings** [comment]
> This looks really clean. Generation has felt like a bit of a second-class citizen in torchtune, and I think this pushes us in the right direction in setting the bar for clean + hackable code across all our recipes. I really like how transparent it makes going from MM prompt -> message, this is probably the most accessible example we have in our codebase.
> 
> I don't have a huge amount to add ontop of Evan's comments. What's the minimum feature set you'd like this to support before it lands?

Minimum feature set is what I have up now. Ideal state would have compile, as well. Quantization will definitely come later.

File: recipes/dev/generate_v2.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (SalmanMohammadi) [reply recipes/dev/generate_v2.py:159]: I guess `generation.generate` would also be used for evals?

[the remark this answers is not in the document]
Comment (joecummings) [reply recipes/dev/generate_v2.py:159]: When I put up the PR, I'd actually love to have a discussion on this. I'm honestly not sure what the best way to balance code readability + abstractions for this kind of thing right now.

[the remark this answers is not in the document]
Comment (SalmanMohammadi) [reply recipes/dev/generate_v2.py:159]: I think this is definitely the way forward for a generation recipe just used to test out your finetuned models, we can benefit from this simplicity by not needing any funky masking/padding logic. It's precise and very easy to understand.

I think it's mostly okay to sacrifice this simplicity when we need to cover broader use-cases and maximise performance like in PPO/evals/agentic etc. How easy it to understand all the compile integrations in the codebase? Not very. The best we can do imo is surface as much of the generation logic as we can so users can reference a single file for all the chunkier generation functionality.

File: recipes/dev/generate_v2.py (hunk 6 above)
Comment (pbontrager) [reply recipes/dev/generate_v2.py:148]: This isn't a pretty solution, but essentially what's happening is during decoding your attention shapes are [1 x kv_cache_size] where 1 is the token being decoded. Inside of TransformerDecoder we take a single slice of a causal mask for our self attention masks. For cross attention, the mask is static during decoding, and the assumption being made here is the decoding tokens have the same mask as the final token of pre-fill e.g. the decoded tokens can see whatever images the previous token could.

File: torchtune/_recipe_registry.py (hunk 1 above)
Comment (pbontrager) [reply torchtune/_recipe_registry.py:256]: I didn't want to call the recipe generate_v2 once it's moved out of dev. But I'm fine with not including dev in the cli name.

File: recipes/configs/dev/multimodal_generation.yaml [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+device: cuda
+dtype: bf16
+seed: 1234
+log_level: INFO
+
+# Generation arguments
+prompt:
+  system: You are a helpful assistant who responds like the author Shakespeare.
+  user:
```
Comment (pbontrager) [inline recipes/configs/dev/multimodal_generation.yaml:32]: Does this support multiple image/text entries or just preprend image?

Comment (joecummings) [reply recipes/configs/dev/multimodal_generation.yaml:32]: prepend

File: recipes/dev/generate_v2.py (hunk 3 above)
Comment (joecummings) [reply recipes/dev/generate_v2.py:30]: I'm being dumb but I'm not sure I understand. Can you show a code sample?

File: recipes/dev/generate_v2.py (hunk 2 above)
Comment (joecummings) [reply recipes/dev/generate_v2.py]: Overkill

**Comment by RdoubleA** [comment]
How is the prompt template applied here if we are not calling `tokenize_messages`?

**Comment by joecummings** [comment]
> How is the prompt template applied here if we are not calling `tokenize_messages`?

We are! The tokenizer / transform __call__ method uses `tokenize_messages` :)

File: recipes/dev/generate_v2.py [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+        model.load_state_dict(_ckpt_dict[training.MODEL_KEY])
+        self.model = model
+        self._logger.info(f"Model was initialized with precision {self._dtype}.")
+
+        # Instantiate transforms
+        self.model_transform = config.instantiate(cfg.transform)
+        self.to_messages = SingleTurnYAMLToMessages()
+
+    @torch.inference_mode()
+    def generate(self, cfg: DictConfig):
```
Comment (SalmanMohammadi) [inline recipes/dev/generate_v2.py:93]: I think this looks pretty neat!

File: torchtune/_recipe_registry.py (hunk 1 above)
Comment (SalmanMohammadi) [reply torchtune/_recipe_registry.py:260]: could move the old recipe into dev as /generate_old

Comment (ebsmothers) [reply torchtune/_recipe_registry.py:260]: Yeah I would just remove dev from the CLI name

File: recipes/dev/generate_v2.py (hunk 3 above)
Comment (ebsmothers) [reply recipes/dev/generate_v2.py]: I mean in practice it's basically the same thing since the signature in `SingleTurnYAMLToMessages` already takes `Dict[str, Any]`. But really you are doing `self.to_messages(cfg.prompt)` and (I think) `cfg.prompt` is an `OmegaConf.DictConfig`. So might be better to just do `self.to_messages(OmegaConf.to_container(cfg.prompt))` so that `SingleTurnYAMLToMessages` is actually just taking a vanilla dict rather than some OmegaConf class.

File: recipes/dev/generate_v2.py [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+            model = config.instantiate(cfg.model)
+        model.load_state_dict(_ckpt_dict[training.MODEL_KEY])
+        self.model = model
+        self._logger.info(f"Model was initialized with precision {self._dtype}.")
+
+        # Instantiate transforms
+        self.model_transform = config.instantiate(cfg.transform)
+        self.to_messages = SingleTurnYAMLToMessages()
+
+    def log_metrics(self, total_time: int, tokens_per_second: float) -> None:
```
Comment (SalmanMohammadi) [inline recipes/dev/generate_v2.py:96]: very nice very hackable

**Comment by SalmanMohammadi** [comment]
hot dog

Is there any reason why this wouldn't work with compile?

File: recipes/dev/generate_v2.py (hunk 11 above)
Comment (joecummings) [reply recipes/dev/generate_v2.py:96]: image

**Comment by joecummings** [comment]
> Yeah I would just remove dev from the CLI name

@ebsmothers Why? I think it makes sense b/c it's buried in dev. Lots of our users use the recipes from the git clone installation.

File: recipes/configs/llama2/generation_v2.yaml [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+#    tune run dev/generate_v2 --config llama2/generation_v2
+
+# Model arguments
+model:
+  _component_: torchtune.models.llama2.llama2_7b
+
+# Transform arguments
+transform:
+  _component_: torchtune.models.llama2.llama2_tokenizer
+  path: /tmp/Llama-2-7b-chat-hf/tokenizer.model
```
[the remark this answers is not in the document]
Comment (joecummings) [reply recipes/configs/llama2/generation_v2.yaml:16]: In reality, I want to move generation to under each model.

**Review by pbontrager — APPROVED** [review]
